src/components/BranchMenu.tsx

```diff
@@ -0,0 +1,315 @@
+   */
+  private _addListeners = () => {
+    // When the repository changes, we're likely to have a new set of branches:
+    this.props.model.repositoryChanged.connect(this._syncState, this);
+
+    // When the HEAD changes, decent probability that we've switched branches:
+    this.props.model.headChanged.connect(this._syncState, this);
+
+    // When the status changes, we may have checked out a new branch (e.g., via the command-line and not via the extension):
+    this.props.model.statusChanged.connect(this._syncState, this);
+  };
+
+  /**
+   * Removes model listeners.
+   */
+  private _removeListeners = () => {
+    this.props.model.repositoryChanged.disconnect(this._syncState, this);
+    this.props.model.headChanged.disconnect(this._syncState, this);
+    this.props.model.statusChanged.disconnect(this._syncState, this);
+  };
+
+  /**
+   * Syncs the component state with the underlying model.
+   */
+  private _syncState = () => {
+    const repo = this.props.model.pathRepository;
+    this.setState({
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    });
+  };
+
+  /**
+   * Callback invoked upon a change to the menu filter.
+   *
+   * @param event - event object
+   */
+  private _onFilterChange = (event: any) => {
+    this.setState({
+      filter: event.target.value
+    });
+  };
+
+  /**
+   * Callback invoked to reset the menu filter.
+   */
+  private _resetFilter = () => {
+    this.setState({
+      filter: ''
+    });
+  };
+
+  /**
+   * Callback invoked upon clicking a button to create a new branch.
+   *
+   * @param event - event object
+   */
+  private _onNewBranchClick = () => {
+    if (!this.props.branching) {
+      showErrorMessage('Creating a new branch is disabled', CHANGES_ERR_MSG);
+      return;
+    }
+    this.setState({
+      branchDialog: true
+    });
+  };
+
+  /**
+   * Callback invoked upon closing a dialog to create a new branch.
+   */
+  private _onNewBranchDialogClose = () => {
+    this.setState({
+      branchDialog: false
+    });
+  };
+
+  /**
+   * Returns a callback which is invoked upon clicking a branch name.
+   *
+   * @param branch - branch name
+   * @returns callback
+   */
+  private _onBranchClickFactory = (branch: string) => {
+    const self = this;
+    return onClick;
+
+    /**
+     * Callback invoked upon clicking a branch name.
+     *
+     * @private
+     * @param event - event object
+     */
+    function onClick() {
+      if (!self.props.branching) {
+        showErrorMessage('Switching branches is disabled', CHANGES_ERR_MSG);
+        return;
+      }
+      const opts = {
+        branchname: branch
+      };
+      self.props.model
+        .checkout(opts)
+        .then(onResolve)
+        .catch(onError);
+    }
+
+    /**
+     * Callback invoked upon promise resolution.
+     *
+     * @private
+     * @param result - result
+     */
+    function onResolve(result: any) {
```

Comment:
Ha you really don't like anonymous functions do you :)

Reply:
@saulshanabrook The standard reason for avoiding anonymous functions (including anonymous function expressions) is for debugging. For example, for callbacks provided to `Promise` methods, stack traces will include the function/method name when the function is named, thus allowing easier mapping of the problematic function to its source implementation (e.g., rather than searching for a line number, one can use `grep`). While benefits may be limited for one-off functions, the benefits increase for functions/methods/design patterns reused across a codebase. 

While named function expression support has improved in recent ECMAScript editions, historically, using them was not "best" practice, due to debugging and browser quirks (e.g., IE).

Lastly, moving the function body to a named function more clearly separates the usage of the function from its implementation and avoids common bracket matching issues (although this is obviated by modern IDE features such as automatic bracket pairing and highlighting, etc). Far too often, developers rely on nested inline anonymous functions which is generally bad practice because often the enclosing scopes include unnecessary identifiers and, without compiler tricks, results in unnecessary stack allocation. Better practice is using named functions at the highest possible scope to both limit a function's scope to what is absolutely necessary (e.g., only those parent scope identifiers which are needed by the function; commonly, this is the top-most module scope) and ensure that JS engines can minimize function creation and GC.

Reply:
Thanks, looking more into this I didn't realize how many ways there were to declare functions in JS...

This is rather surprising to me, coming from a Python world, but it looks like if you set a variable to an arrow function it will get its name from the variable name (https://stackoverflow.com/a/37488652). 

> Far too often, developers rely on nested inline anonymous functions which is generally bad practice because often the enclosing scopes include unnecessary identifiers and, without compiler tricks, results in unnecessary stack allocation. Better practice is using named functions at the highest possible scope to both limit a function's scope to what is absolutely necessary (e.g., only those parent scope identifiers which are needed by the function; commonly, this is the top-most module scope)

Is there a difference between arrow function scoping and regular function scoping? I know that arrow functions don't have a new `this` so you don't have to have the `self = this` line with them.

Reply:
@saulshanabrook Re: anonymous function expression assignments. Correct, but this is a recent feature stemming from ES6/ES2015. Historically, this was not the case. In stack traces, the function would be marked as "anonymous" despite assignment to a variable.

Re: scoping. None beside lexical `this`. Scoping still works the same. A nested arrow function will still be able to access variables from the parent scope.